### PR TITLE
feat: CRP-2611 Introduce new context type for VetKD requests

### PR DESCRIPTION
**Pull request by eichhorl**
We extend the existing `SignWithThresholdContext` by adding another `ThresholdArguments` variant of type `VetKdArguments`. Contexts of this type cannot be created yet.

In addition, the compiler forces us to define a cycles cost for vet KD requests. For now we choose the same cost as for tEcdsa and tSchnorr, but this may be refined further at a later stage.

There are some parts in consensus where we only want to handle tSchnorr and tEcdsa contexts (as they rely on IDKG), but ignore vet KD contexts (as they rely on NiDKG). For that reason we add a new type `IDkgSignWithThresholdContext` which wraps the reference to a `SignWithThresholdContext`, but only if that context may be handled by IDKG. Functions in the IDKG component are then changed to accept the new type.

### CODE REVIEW

File: rs/replicated_state/src/metadata_state/tests.rs [hunk 1]
```diff
@@ -2298,7 +2298,7 @@ fn compatibility_for_cycles_use_case() {
         CyclesUseCase::iter()
             .map(|x| x as i32)
             .collect::<Vec<i32>>(),
-        [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]
+        [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]
```
Comment (alin-at-dfinity) [inline rs/replicated_state/src/metadata_state/tests.rs:2301]: FYI, what this is saying (among other things) is that you should not be immediately be producing this new variant in the next replica release, because if we need to revert to the previous replica version and this variant is in any canister's state, we will panic while reading the checkpoint.

I haven't looked at the rest of the change, so I don't know if the PR actually makes it possible for this new use case to be charged for. If so, then you will first need to make the replicated state change and then, in a subsequent replica release, merge the rest of your PR. If not, then you just need to wait until a subsequent replica release to enable the functionality.

Comment (eichhorl) [reply rs/replicated_state/src/metadata_state/tests.rs:2301]: Thanks for looking out! Yes, if I understand correctly, the new variant will only be produced if we call `sign_with_threshold` with the new VetKd arguments. This isn't implemented yet, and even once it is, that function will only be called once we generate a new key on mainnet and enable the feature via proposal.

File: rs/replicated_state/src/metadata_state/subnet_call_context_manager.rs [hunk 2]
```diff
@@ -901,6 +1001,16 @@ impl SignWithThresholdContext {
             _ => panic!("Schnorr arguments not found."),
         }
     }
+
+    /// Returns Vetkdguments.
```
Comment (eichhorl) [inline rs/replicated_state/src/metadata_state/subnet_call_context_manager.rs:1005]:
```suggestion
    /// Returns VetKd arguments.
```

File: rs/replicated_state/src/metadata_state/tests.rs (hunk 1 above)
Comment (alin-at-dfinity) [reply rs/replicated_state/src/metadata_state/tests.rs:2301]: Then this all LGTM.

I'll leave the approval of `rs/replicated_state` changes to someone from Execution, as they own `SubnetCallContextManager`.

File: rs/protobuf/def/state/metadata/v1/metadata.proto [hunk 3]
```diff
@@ -78,10 +78,19 @@ message SchnorrArguments {
   bytes message = 2;
 }
 
+message VetKdArguments {
+  types.v1.VetKdKeyId key_id = 1;
+  bytes derivation_id = 2;
+  bytes encryption_key = 3;
```
Comment (fspreiss) [inline rs/protobuf/def/state/metadata/v1/metadata.proto:84]: nit: How about `encryption_public_key` (see my other comment)?

**Review by maksymar — APPROVED** [review]

**Review by fspreiss — APPROVED** [review]
Thanks, @eichhorl! LGTM!

File: rs/consensus/src/idkg/signer.rs [hunk 4]
```diff
@@ -551,11 +553,17 @@ impl ThresholdSigner for ThresholdSignerImpl {
             .get_state()
             .signature_request_contexts()
             .iter()
-            .flat_map(|(callback_id, context)| {
-                context.matched_pre_signature.map(|(_, height)| RequestId {
+            .flat_map(|(callback_id, context)| match &context.args {
+                ThresholdArguments::Ecdsa(_) | ThresholdArguments::Schnorr(_) => {
```
Comment (Sawchord) [inline rs/consensus/src/idkg/signer.rs:557]: Maybe also implement `is_idkg()` and `is_vetkd()` functions for `TresholdArguments`? Makes for cleaner code

File: rs/consensus/src/idkg/test_utils.rs [hunk 5]
```diff
@@ -97,7 +103,13 @@ fn fake_signature_request_args(key_id: IDkgMasterPublicKeyId) -> ThresholdArgume
             key_id,
             message: Arc::new(vec![1; 48]),
         }),
-        MasterPublicKeyId::VetKd(_) => panic!("not applicable to vetKD"),
+        MasterPublicKeyId::VetKd(key_id) => ThresholdArguments::VetKd(VetKdArguments {
```
Comment (Sawchord) [inline rs/consensus/src/idkg/test_utils.rs:106]: This should probably be lifted out of the `idkg` submodule, but we can do this in a separate PR.

This likely applies to the other changes of `test_utils` as well.

File: rs/consensus/src/idkg/payload_verifier.rs [hunk 6]
```diff
@@ -1143,6 +1149,55 @@ mod test {
         );
     }
 
+    #[test]
```
Comment (Sawchord) [inline rs/consensus/src/idkg/payload_verifier.rs:1152]: I'm not sure I understand this test fully.

If I understand correctly, we are testing that `validate_new_signature_agreements` ignores `signature_request_contexts` of vetkd requests, but since we can not directly check that, we insert a vetkd context with `pseudo_random_id` and then check that the function does not return that contextr (since it only returns idkg contexts).

If this is the case, I think it might sense to describe this a little bit better, because it is not so obvious.

Also, we might want to rename `signature_request_contexts` to `idkg_signature_request_contexts`, to drive the point home

**Review by Sawchord — APPROVED** [review]
LGTM

**Review by fspreiss — APPROVED** [review]

**Review by maksymar — APPROVED** [review]

File: rs/replicated_state/src/metadata_state/subnet_call_context_manager.rs [hunk 7]
```diff
@@ -802,10 +805,45 @@ impl TryFrom<pb_metadata::SchnorrArguments> for SchnorrArguments {
     }
 }
 
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct VetKdArguments {
+    pub key_id: VetKdKeyId,
+    pub derivation_id: Vec<u8>,
+    pub encryption_key: Vec<u8>,
```
[the remark this answers is not in the document]
Comment (fspreiss) [reply rs/replicated_state/src/metadata_state/subnet_call_context_manager.rs]: > In the VetKdArgs, we called it encryption_key, but we probably should rename it there too (out of scope of this PR here).

Done in https://github.com/dfinity/ic/pull/2789
